compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -150,6 +150,29 @@ unknown unitId pkgMap = conversionError errMsg
 unhandled :: (HasCallStack, Data a, Outputable a) => String -> a -> ConvertM e
 unhandled typ x = unsupported (typ ++ " with " ++ lower (show (toConstr x))) x
 
+warn :: String -> ConvertM ()
+warn msg = do
+    ConversionEnv{..} <- ask
+    let diag = (convModuleFilePath, ShowDiag, Diagnostic
+            { _range = maybe noRange sourceLocToRange convRange
+            , _severity = Just DsWarning
+            , _source = Just "Core to DAML-LF"
+            , _message = T.pack msg
+            , _code = Nothing
+            , _relatedInformation = Nothing
+            , _tags = Nothing
+            })
+    modify' $ \st -> ConversionState (diag : convWarnings st)
+
+warnNotDataDependable :: String -> ConvertM ()
+warnNotDataDependable what =
+    warn $ unlines
+        [ "Using " ++ what ++ " in combination with type classes"
```

Comment:
Is this really only a problem in combination with typeclasses? I thought erased kinds cause problems everywhere.

Reply:
You're right, erased kinds are problematic everywhere. With code using promoted kinds does not compile in the first case. Functions using promoted kinds in their signature will have a different signature after importing via data-dependencies. This might become annoying but does not stop a package from being imported. The problem with typeclasses is that you might end up with overlapping instances after kind erasure where there was no overlap before. This will prevent a package from being data-dependable. I still don't understand the full picture and I expect these warnings to extend to more places in the long run.